nni/nas/space/space.py

```diff
@@ -0,0 +1,388 @@
+            exception = self.evaluator.check_contains(sample)
+            if exception is not None:
+                exception.paths.append('evaluator')
+                return exception
+        return None
+
+    def executable_model(self) -> Any:
+        return self.model_space
+
+    def leaf_mutables(self, is_leaf: Callable[[Mutable], bool]) -> Iterable[LabeledMutable]:
+        yield from self.model_space.leaf_mutables(is_leaf)
+        if isinstance(self.evaluator, Mutable):
+            yield from self.evaluator.leaf_mutables(is_leaf)
+
+    def _dump(self) -> NoReturn:
+        """Serialization of a :class:`RawFormatModelSpace` is not supported.
+
+        Notes
+        -----
+        The potential issues with serialization are in two folds:
+
+        1. The model space could be a deep learning model, and have been arbitrarily mutated by the strategy (e.g., one-shot).
+           For example, one submodule is replaced by another, or a layer is removed.
+           In this case, we surely cannot use the init arguments to recover the model.
+        2. The model space could contain parameters (weights), that are meant to be part of the model space.
+           (That's why we have :class:`RawFormatModelSpace` other than :class:`SimplifiedModelSpace`).
+           In this case, we need to dump all the parameters, which could be potentially large and slow.
+
+        One potential solution to this problem might be introducing an advanced version of ``nni.trace``,
+        that allows users / strategies to define a function that recreates the current instance from scratch.
+        This function must be runnable in a completely new isolated process.
+
+        Another potential solution might be introducing several flags to evaluator to specific needs like one-shot.
+        But I don't think it's a good idea, because I want to make the evaluator semantically simple.
+        """
+        raise NotImplementedError('`_dump` is not implemented for {}'.format(self.__class__.__name__))
+
+    @classmethod
+    def _load(cls, **kwargs) -> NoReturn:
+        raise NotImplementedError('`_load` is not implemented for {}'.format(RawFormatModelSpace.__name__))
+
+
+class SimplifiedModelSpace(ExecutableModelSpace):
+    """Model space that is simplified (see :meth:`~nni.mutable.Mutable.simplify`),
+    and only keeps the key information.
+
+    With :class:`SimplifiedModelSpace`, all details inside the model will be removed,
+    which means, the weights, attributes, inplace modifications of the model will all be lost.
+    Only the simplified mutables and necessary init arguments to recover the model for execution will be kept.
+
+    To work with :class:`SimplifiedModelSpace`,
+    the model itself should detect :meth:`~nni.nas.space.current_model` context,
+    and init a sampled concrete model when it's inside the context.
+    Since the model will be recreated, ``freeze`` and ``contains`` method of model space is never used.
```

Comment:
why do we have to introduce "context" in `SimplifiedModeSpace`?

Reply:
I've rewritten the paragraph to make it somewhat more comprehensible. Might be more clearly explained with #5279.